src/movies.js

```diff
@@ -1,16 +1,150 @@
 
-// Iteration 2: Steven Spielberg. The best? - How many drama movies did STEVEN SPIELBERG direct?
+function howManyMovies(array) {
+  let spielbergMovie = array.filter((movie) => {
+    return (
+      movie.director === "Steven Spielberg" &&
+      movie.genre.indexOf("Drama") != -1
+    );
+  });
+  return spielbergMovie.length;
+}
 
 // Iteration 3: All rates average - Get the average of all rates with 2 decimals
 
+function ratesAverage(array) {
+  if (array.length !== 0) {
+    let total = array.reduce((acc, movie) => {
+      if (movie.rate != "" && movie.rate != undefined) {
+        return acc + movie.rate;
+      } else {
+        return acc;
+      }
+    }, 0);
+    total = total / array.length;
+    return Number(total.toFixed(2));
+  } else {
+    return 0;
+  }
+}
+
 // Iteration 4: Drama movies - Get the average of Drama Movies
 
-// Iteration 5: Ordering by year - Order by year, ascending (in growing order)
+function dramaMoviesRate(array) {
+  let dramaMovies = array.filter((movie) => {
+    return (
+      movie.rate != "" &&
+      movie.rate != undefined &&
+      movie.genre.indexOf("Drama") != -1
+    );
+  });
+
+  if (dramaMovies.length !== 0) {
+    let total = dramaMovies.reduce((acc, movie) => {
+      if (dramaMovies.length > 0) {
+        return acc + movie.rate;
+      } else {
+        return acc;
+      }
+    }, 0);
+    total = total / dramaMovies.length;
+    return Number(total.toFixed(2));
+  } else {
+    return 0;
+  }
+}
+
+// // Iteration 5: Ordering by year - Order by year, ascending (in growing order)
+
+function orderByYear(arr) {
+  let newMovies = JSON.parse(JSON.stringify(arr));
+
+  newMovies.sort((a, b) => {
+    if (a.year > b.year) {
+      return 1;
+    } else if (a.year < b.year) {
+      return -1;
+    } else {
+      if (a.title > b.title) {
+        return 1;
+      } else if (a.title < b.title) {
+        return -1;
+      } else {
+        return 0;
+      }
+    }
+  });
+  return newMovies;
+}
+
+// // Iteration 6: Alphabetic Order - Order by title and print the first 20 titles
+
+function orderAlphabetically(arr) {
+  let newMovies = JSON.parse(JSON.stringify(arr));
+
+  let movieTitles = newMovies
+    .sort((a, b) => {
+      if (a.title > b.title) {
+        return 1;
+      } else if (a.title < b.title) {
+        return -1;
+      } else {
+        return 0;
+      }
+    })
+    .map((movie) => {
+      return movie.title;
+    });
+
+  if (movieTitles.length > 21) {
+    movieTitles = movieTitles.splice(0, 20);
+    return movieTitles;
+  } else {
+    return movieTitles;
+  }
+}
+
+// // BONUS - Iteration 7: Time Format - Turn duration of the movies from hours to minutes
+function turnHoursToMinutes(array) {
+  let newMovies = array.map((movie) => {
+    if (
+      movie.duration.indexOf("h") !== -1 &&
+      movie.duration.indexOf("min") === -1
+    ) {
+      let hour = Number(movie.duration[0] * 60);
+      movie.duration = hour;
```

Comment:
even though each movie is being passed through the parameter of the map, this still points to the original reference, so doing movie.duration will modify the original data. Doing a deep clone either of the array at the start of the function or of each movie obj at the start of the .map will solve this.